File: app/src/main/java/com/mmjang/ankihelper/ui/plan/PlansAdapter.java
package com.mmjang.ankihelper.ui.plan;

import android.app.Activity;
import android.content.DialogInterface;
import android.content.Intent;
import android.graphics.Color;
import androidx.appcompat.app.AlertDialog;
import androidx.recyclerview.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;
import android.widget.RelativeLayout;
import android.widget.TextView;

import com.mmjang.ankihelper.MyApplication;
import com.mmjang.ankihelper.R;
import com.mmjang.ankihelper.data.database.ExternalDatabase;
import com.mmjang.ankihelper.data.plan.OutputPlan;
import com.mmjang.ankihelper.data.plan.OutputPlanPOJO;
import com.mmjang.ankihelper.ui.plan.helper.ItemTouchHelperAdapter;
import com.mmjang.ankihelper.ui.plan.helper.ItemTouchHelperViewHolder;
import com.mmjang.ankihelper.util.DialogUtil;
import com.mmjang.ankihelper.util.Utils;

import org.litepal.crud.LitePalSupport;

import java.util.List;


/**
 * Created by liao on 2017/4/27.
 */

public class PlansAdapter extends RecyclerView.Adapter<PlansAdapter.ViewHolder> implements ItemTouchHelperAdapter{
    private List<OutputPlanPOJO> mPlansList;
    private Activity mActivity;

    static class ViewHolder extends RecyclerView.ViewHolder implements ItemTouchHelperViewHolder{
        RelativeLayout countainer;
        TextView planName;
        TextView dictName;
        LinearLayout layoutEdit;
        LinearLayout layoutDelete;

        public ViewHolder(View view) {
            super(view);
            countainer = view.findViewById(R.id.plan_item);
            planName = (TextView) view.findViewById(R.id.plans_name);
            dictName = (TextView) view.findViewById(R.id.plans_dictionary_name);
            layoutEdit = (LinearLayout) view.findViewById(R.id.layout_edit);
            layoutDelete = (LinearLayout) view.findViewById(R.id.layout_delete);
        }

        @Override
        public void onItemSelected() {
            countainer.setBackgroundColor(Color.LTGRAY);
        }

        @Override
        public void onItemClear() {
            countainer.setBackgroundColor(0);
        }
    }

    public PlansAdapter(
            Activity activity,
            List<OutputPlanPOJO> planList) {
        mPlansList = planList;
        mActivity = activity;
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(parent.getContext())
                .inflate(R.layout.item_plans, parent, false);
        ViewHolder holder = new ViewHolder(view);
        return holder;
    }

    @Override
    public void onBindViewHolder(final ViewHolder holder, final int position) {
        OutputPlanPOJO plan = mPlansList.get(position);
        holder.planName.setText(plan.getPlanName());
        holder.dictName.setText(plan.getDictionaryKey());
        holder.layoutDelete.setOnClickListener(
                new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {

                        new AlertDialog.Builder(mActivity)
                                .setTitle(R.string.confirm_deletion)
                                //.setMessage("Do you really want to whatever?")
                                .setIcon(android.R.drawable.ic_dialog_alert)
                                .setPositiveButton(android.R.string.yes, new DialogInterface.OnClickListener() {
                                    public void onClick(DialogInterface dialog, int whichButton) {
                                        int pos = holder.getAdapterPosition();
                                        //mPlansList.get(pos).delete();
                                        ExternalDatabase.getInstance().deletePlanByName(mPlansList.get(pos).getPlanName());
                                        mPlansList.remove(pos);
                                        notifyItemRemoved(pos);
                                    }
                                })
                                .setNegativeButton(android.R.string.no, null).show();
                    }
                }
        );

        holder.layoutEdit.setOnClickListener(
                new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        if (MyApplication.getAnkiDroid(MyApplication.getContext()).isAnkiDroidRunning()) {
                            int pos = holder.getAdapterPosition();
                            String planName = mPlansList.get(pos).getPlanName();
                            Intent intent = new Intent(mActivity, PlanEditorActivity.class);
                            intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
                            intent.setAction(Intent.ACTION_SEND);
                            intent.putExtra(Intent.EXTRA_TEXT, planName);
                            MyApplication.getContext().startActivity(intent);
                        } else {
                            DialogUtil.showStartAnkiDialog(mActivity);
                        }
                    }
                }
        );

    }

    @Override
    public void onItemMove(int fromPosition, int toPosition) {
        OutputPlanPOJO from = mPlansList.get(fromPosition);
        mPlansList.remove(fromPosition);
        mPlansList.add(toPosition, from);
        notifyItemMoved(fromPosition, toPosition);
    }

    @Override
    public void onMoveFinished() {
        ExternalDatabase.getInstance().refreshPlanWith(mPlansList);
//        List<OutputPlan> plansInDatabase = DataSupport.findAll(OutputPlan.class);
//        for(int i = 0; i < plansInDatabase.size(); i ++){
//            OutputPlan oldPlan = plansInDatabase.get(i);
//            OutputPlan newPlan = mPlansList.get(i);
//            if(oldPlan.getPlanName() == newPlan.getPlanName()){
//                continue;
//            }else{
//                oldPlan.setPlanName(newPlan.getPlanName());
//                oldPlan.setDictionaryKey(newPlan.getDictionaryKey());
//                oldPlan.setOutputDeckId(newPlan.getOutputDeckId());
//                oldPlan.setOutputModelId(newPlan.getOutputModelId());
//                oldPlan.setFieldsMap(newPlan.getFieldsMap());
//                oldPlan.save();
//            }
//        }
    }

    @Override
    public int getItemCount() {
        return mPlansList.size();
    }
}

File: app/src/main/java/com/lmyby/ankiquicker/ui/plan/PlansAdapter.kt
package com.lmyby.ankiquicker.ui.plan

import android.app.Activity
import android.content.Intent
import android.graphics.Color
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout
import android.widget.RelativeLayout
import android.widget.TextView
import androidx.appcompat.app.AlertDialog
import androidx.lifecycle.LifecycleOwner
import androidx.recyclerview.widget.RecyclerView
import com.lmyby.ankiquicker.MyApplication
import com.lmyby.ankiquicker.R
import com.lmyby.ankiquicker.data.database.AppDatabase
import com.lmyby.ankiquicker.data.plan.OutputPlanEntity
import com.lmyby.ankiquicker.data.plan.OutputPlanPOJO
import com.lmyby.ankiquicker.data.plan.OutputPlanRepository
import com.lmyby.ankiquicker.data.plan.OutputPlanRepositoryHelper
import com.lmyby.ankiquicker.ui.plan.helper.ItemTouchHelperAdapter
import com.lmyby.ankiquicker.ui.plan.helper.ItemTouchHelperViewHolder
import com.lmyby.ankiquicker.util.DialogUtil

/**
 * Adapter for output plans list with drag-and-drop support
 * Converted to Kotlin as part of Phase 4 adapter migration
 */
class PlansAdapter(
    private val mActivity: Activity,
    private val mPlansList: MutableList<OutputPlanPOJO>
) : RecyclerView.Adapter<PlansAdapter.ViewHolder>(), ItemTouchHelperAdapter {

    private val repositoryHelper: OutputPlanRepositoryHelper

    init {
        // Initialize repository helper
        val database = AppDatabase.getInstance(mActivity.applicationContext)
        val repository = OutputPlanRepository(database.outputPlanDao())
        repositoryHelper = OutputPlanRepositoryHelper(repository, mActivity as LifecycleOwner)
    }

    class ViewHolder(view: View) : RecyclerView.ViewHolder(view), ItemTouchHelperViewHolder {
        val container: RelativeLayout = view.findViewById(R.id.plan_item)
        val planName: TextView = view.findViewById(R.id.plans_name)
        val dictName: TextView = view.findViewById(R.id.plans_dictionary_name)
        val layoutEdit: LinearLayout = view.findViewById(R.id.layout_edit)
        val layoutDelete: LinearLayout = view.findViewById(R.id.layout_delete)

        override fun onItemSelected() {
            container.setBackgroundColor(Color.LTGRAY)
        }

        override fun onItemClear() {
            container.setBackgroundColor(0)
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
        val view = LayoutInflater.from(parent.context)
            .inflate(R.layout.item_plans, parent, false)
        return ViewHolder(view)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val plan = mPlansList[position]
        holder.planName.text = plan.planName
        holder.dictName.text = plan.dictionaryKey

        holder.layoutDelete.setOnClickListener {
            AlertDialog.Builder(mActivity)
                .setTitle(R.string.confirm_deletion)
                .setIcon(android.R.drawable.ic_dialog_alert)
                .setPositiveButton(android.R.string.yes) { _, _ ->
                    val pos = holder.bindingAdapterPosition
                    val planName = mPlansList[pos].planName

                    // Delete plan using repository (fire-and-forget)
                    repositoryHelper.deletePlanFireAndForget(planName)

                    mPlansList.removeAt(pos)
                    notifyItemRemoved(pos)
                }
                .setNegativeButton(android.R.string.no, null)
                .show()
        }

        holder.layoutEdit.setOnClickListener {
            if (MyApplication.getAnkiDroid(MyApplication.getContext()).isAnkiDroidRunning) {
                val pos = holder.bindingAdapterPosition
                val planName = mPlansList[pos].planName
                val intent = Intent(mActivity, PlanEditorActivity::class.java).apply {
                    flags = Intent.FLAG_ACTIVITY_NEW_TASK
                    action = Intent.ACTION_SEND
                    putExtra(Intent.EXTRA_TEXT, planName)
                }
                MyApplication.getContext().startActivity(intent)
            } else {
                DialogUtil.showStartAnkiDialog(mActivity)
            }
        }
    }

    override fun onItemMove(fromPosition: Int, toPosition: Int) {
        val from = mPlansList.removeAt(fromPosition)
        mPlansList.add(toPosition, from)
        notifyItemMoved(fromPosition, toPosition)
    }

    override fun onMoveFinished() {
        // Convert POJOs to Entities
        val entities = mPlansList.map { pojo ->
            OutputPlanEntity().apply {
                planName = pojo.planName
                dictionaryKey = pojo.dictionaryKey
                outputDeckId = pojo.outputDeckId
                outputModelId = pojo.outputModelId
                fieldsMap = pojo.getFieldsMapString()
            }
        }

        // Refresh all plans using repository (fire-and-forget)
        repositoryHelper.refreshAllPlansFireAndForget(entities)
    }

    override fun getItemCount(): Int = mPlansList.size
}
